src/dialects/abstract/query-generator-typescript.ts

```diff
@@ -0,0 +1,129 @@
+    }
+
+    if (!options.dialect) {
+      throw new Error('QueryGenerator initialized without options.dialect');
+    }
+
+    this.sequelize = options.sequelize;
+    this.dialect = options.dialect;
+  }
+
+  protected get options() {
+    return this.sequelize.options;
+  }
+
+  extractTableDetails(
+    tableNameOrModel: TableNameOrModel,
+    options?: { schema?: string, delimiter?: string },
+  ): TableNameWithSchema {
+    const tableNameObject = isModelStatic(tableNameOrModel) ? tableNameOrModel.getTableName()
+      : isString(tableNameOrModel) ? { tableName: tableNameOrModel }
+      : tableNameOrModel;
+
+    if (!isPlainObject(tableNameObject)) {
+      throw new Error(`Invalid input received, got ${NodeUtil.inspect(tableNameOrModel)}, expected a Model Class, a TableNameWithSchema object, or a table name string`);
+    }
+
+    return {
+      ...tableNameObject,
+      schema: options?.schema || tableNameObject.schema || this.options.schema || this.dialect.getDefaultSchema(),
+      delimiter: options?.delimiter || tableNameObject.delimiter || '.',
+    };
+  }
+
+  /**
+   * Quote table name with optional alias and schema attribution
+   *
+   * @param param table string or object
+   * @param alias alias name
+   */
+  quoteTable(param: TableNameOrModel, alias: boolean | string = false): string {
+    if (isModelStatic(param)) {
+      param = param.getTableName();
+    }
+
+    const tableName = this.extractTableDetails(param);
+
+    if (isObject(param) && ('as' in param || 'name' in param)) {
+      throw new Error('parameters "as" and "name" are not allowed in the first parameter of quoteTable, pass them as the second parameter.');
+    }
+
+    if (alias === true) {
+      alias = tableName.tableName;
+    }
+
+    let sql = '';
+
+    if (this.dialect.supports.schemas) {
+      // Some users sync the same set of tables in different schemas for various reasons
+      // They then set `searchPath` when running a query to use different schemas.
+      // See https://github.com/sequelize/sequelize/pull/15274#discussion_r1020770364
+      // For this reason, we treat the default schema as equivalent to "no schema specified"
+      if (tableName.schema && tableName.schema !== this.dialect.getDefaultSchema()) {
+        sql += `${this.quoteIdentifier(tableName.schema)}.`;
+      }
+
+      sql += this.quoteIdentifier(tableName.tableName);
+    } else {
+      const fakeSchemaPrefix = (tableName.schema && tableName.schema !== this.dialect.getDefaultSchema())
+        ? tableName.schema + (tableName.delimiter || '.')
+        : '';
+
+      sql += this.quoteIdentifier(fakeSchemaPrefix + tableName.tableName);
+    }
+
+    if (alias) {
+      sql += ` AS ${this.quoteIdentifier(alias)}`;
+    }
+
+    return sql;
+  }
+
+  /**
+   * Adds quotes to identifier
+   *
+   * @param identifier
+   * @param _force
+   */
+  quoteIdentifier(identifier: string, _force?: boolean) {
```

Comment:
Do we have unit tests for these?

Reply:
A single one in `mssql/query-generator.test.js`. I've moved it to our common query-generator tests